### ISSUE TITLE
"error: nothing to do" for the past week

### ISSUE THREAD

**Issue by boxleytw** [Homebrew/homebrew-core#52623]
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

The last two checkboxes are not relevant (I was updating Homebrew, not installing formulas.)

## What you were trying to do (and why)

Running `brew update` for obvious reasons.  This has been happening for roughly the last week.  I've been using Homebrew for years, and never seen this behavior before.  I'm rather suspicious that no packages have updated in around 10 days.

## What happened (include command output)

```
🌼 brew update
error: nothing to do
error: nothing to do
error: nothing to do
Already up-to-date.
```

```
🌼 PATH=/usr/bin:"$PATH" brew update
error: nothing to do
error: nothing to do
error: nothing to do
Already up-to-date.
```

## What you expected to happen

See updates, and not see "error" messages.

## Step-by-step reproduction instructions (by running `brew install` commands)

Run `brew update`.

## More details

```
🌼 brew config
HOMEBREW_VERSION: 2.2.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: fa8fe0fc396550c23156043cd07550fc6ae6a27a
Last commit: 13 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e81f13c3917e21fdbc2e905e90c0ef5fd643c80b
Core tap last commit: 2 weeks ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_VISUAL: vim
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
Clang: 11.0 build 1103
Git: 2.26.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 14, 11.0.6
macOS: 10.15.3-x86_64
CLT: 11.4.0.0.1.1583798712
Xcode: 11.4
```

```
🌼 brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed coreutils in your path can cause gmp builds to fail.

Warning: Putting non-prefixed findutils in your path can cause python builds to fail.
(Exit 1) brew doctor
```

Note: the last line is from my shell, not Homebrew; I have my Bash prompt configured to notify me when a command exits non-zero.

```
🌼 brew list
ascii			jenv			openssl@1.1
bash			jq			pandoc
bash-completion		kotlin			pcre2
coreutils		kscript			perl
cowsay			ldns			pkg-config
curl			libcbor			python
expect			libfido2		python@3.8
findutils		libssh2			readline
fortune			libyaml			rlwrap
gdbm			lolcat			ruby
gettext			lua			rustc-completion
git			lynx			sqlite
go			make			thefuck
golangci-lint		node			vim
httpie			oniguruma		xz
icu4c			openssh
```

```
🌼 brew cask list
visual-studio-code
```

I tried diagnosing this more myself, to see if there were some `git` issue specific to me, but was concerned with fiddling too much with internals of Homebrew:

```
🌼 cd /usr/local/Homebrew
🌼 git pull
There is no tracking information for the current branch.
Please specify which branch you want to rebase against.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> stable

(Exit 1) git pull
```

**Comment by Bo98** [comment]
Does `brew update-reset` fix things?

**Comment by boxleytw** [comment]
@Bo98 Wow!  I never knew about that command.

I no longer get the "error: nothing to do" messages, thank you!  However, there are no updates ... yet.

```
🌼 brew update
Already up-to-date.
```

I'll try updating throughout the week, and see if updates come along.

Is there a straight-forward way for me, using a browser, to check on the packages I have installed?  I'd like to confirm that if command line is matching reality.

**Comment by Bo98** [comment]
> I'd like to confirm that the command line is matching reality

Probably the best way is to see "Core tap last commit" in `brew config`. It should typically be in the range of hours or minutes, not weeks.

> It might be a nice enhancement for Homebrew to self-diagnose when encountering "error: nothing to do"

Probably, though it's the first time I've seen it happen. I'll see if can find a specific cause.

**Comment by boxleytw** [comment]
@Bo98 Awesome, thank you again.  Last "Core tap commit" was two hours ago.  I'll close this issue.

```
🌼 brew config
HOMEBREW_VERSION: 2.2.11-136-g842d92a
ORIGIN: https://github.com/Homebrew/brew
HEAD: 842d92a6c5ae3530738ad4ba09a5a00d89fd34bc
Last commit: 4 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 8b2b4d60b802b40623484096d7e810982d272a80
Core tap last commit: 2 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_VISUAL: vim
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
Clang: 11.0 build 1103
Git: 2.26.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 14, 11.0.6
macOS: 10.15.3-x86_64
CLT: 11.4.0.0.1.1583798712
Xcode: 11.4
```

**Comment by SMillerDev** [comment]
By the way, `brew outdated` will show you what is outdated (since `update-reset` won't)

**Comment by boxleytw** [comment]
Oh, @SMillerDev that is interesting.  `brew update` says nothing to update.  But `brew outdated` says:

```
🌼 brew outdated
go (1.14) < 1.14.1
kotlin (1.3.70) < 1.3.71
node (13.11.0) < 13.12.0
openssl@1.1 (1.1.1d) < 1.1.1f
pandoc (2.9.2) < 2.9.2.1
pkg-config (0.29.2) < 0.29.2_1
ruby (2.7.0) < 2.7.1
vim (8.2.0400) < 8.2.0500
```

**Comment by SMillerDev** [comment]
That's the correct behavior. `brew update` will only tell you what changed since the last time you ran it. `brew upgrade` will actually upgrade the things that are outdated. As for the `error: nothing to do`, that's better debugged over at `Homebrew/brew`

**Comment by boxleytw** [comment]
@SMillerDev Hmmm ... the user experience of seeing "error" when nothing is wrong is awkward.  And it feels strange as well for `brew update` to show no work followed by `brew upgrade` which does, in fact, upgrade packages.

**Comment by Bo98** [comment]
The error indeed shouldn't happen. I don't know what causes it though.

`brew update` and `brew upgrade` do different things. The former updates Homebrew itself and the formulae files. `brew upgrade` upgrades installed packages. (https://docs.brew.sh/FAQ#how-do-i-update-my-local-packages)

**Comment by SMillerDev** [comment]
That's actually pretty standard behavior for package managers. Apt does the same thing for example. You don't always want to update everything as soon as it's available.

I'm going to close this issue since it doesn't belong here. You could make a new one in https://github.com/Homebrew/brew for the reported error if no such issue exists yet.

**Comment by boxleytw** [comment]
@SMillerDev Thanks for your help here.

I'm using `brew update-reset` to force finding new packages, followed with `brew upgrade`, since `brew update` no longer prints package updates.

**Comment by MikeMcQuaid** [comment]
https://github.com/Homebrew/brew/issues/7374 may be related (h/t @bo98)

**Comment by boxleytw** [comment]
I found this finally resolved by version 2.26.1_1 of git.

**Comment by MikeMcQuaid** [comment]
Fixed by https://github.com/Homebrew/homebrew-core/pull/53266
